core/trino-main/src/main/java/io/trino/execution/QueryManagerConfig.java

```diff
@@ -55,7 +55,7 @@ public class QueryManagerConfig
     public static final long AVAILABLE_HEAP_MEMORY = Runtime.getRuntime().maxMemory();
     public static final int MAX_TASK_RETRY_ATTEMPTS = 126;
     public static final int FAULT_TOLERANT_EXECUTION_MAX_PARTITION_COUNT_LIMIT = 1000;
-
+    public static final int DISPATCHER_THREADPOOL_MAX_SIZE = Runtime.getRuntime().availableProcessors() * 10;
```

Comment:
If IO are performed then using `Runtime.getRuntime().availableProcessors()` doesn't really make much sense as these threads are not really CPU intensive. In fact, I think `Runtime.getRuntime().availableProcessors() * 10` will hurt concurrent scenarios.

Why can't you limit concurrency with resource queue limit rather than adding global thread cap?

Keep in mind that these queries will continue to queue up in dispatcher executor and will eventually fail in some weird way (and probably starve some other queries in the process). It's better to fail them gracefully via resource group

Reply:
> If IO are performed then using `Runtime.getRuntime().availableProcessors()` doesn't really make much sense as these threads are not really CPU intensive. In fact, I think `Runtime.getRuntime().availableProcessors() * 10` will hurt concurrent scenarios.
> 
> Why can't you limit concurrency with resource queue limit rather than adding global thread cap?
> 
> Keep in mind that these queries will continue to queue up in dispatcher executor and will eventually fail in some weird way (and probably starve some other queries in the process). It's better to fail them gracefully via resource group

thanks @sopel39 for reviewing!

Please guide us, to protect coordinator in this kind of case.

I might be wrong, but with the resource group, we have seen that even though there were max queue defined, it was not blocking the queries.

I need to double check, but I think the applying resource group step is after we use the `dispatchExecutor` here : https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/dispatcher/DispatchManager.java#L159 
....

there is another interesting codepath I found:  https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQueryFactory.java#L146

It is also using the same  

> If IO are performed then using `Runtime.getRuntime().availableProcessors()` doesn't really make much sense as these threads are not really CPU intensive. In fact, I think `Runtime.getRuntime().availableProcessors() * 10` will hurt concurrent scenarios.
> 
> Why can't you limit concurrency with resource queue limit rather than adding global thread cap?
> 
> Keep in mind that these queries will continue to queue up in dispatcher executor and will eventually fail in some weird way (and probably starve some other queries in the process). It's better to fail them gracefully via resource group

@sopel39 thanks for the review!

I am not be aware of the overall code flow here, so seeking guidance.
it would be great, if we can handle it using the resource group.

on the resource group path: I might be wrong, when I followed the code, I was thinking does resource group is applied after the `dispatchExecutor` being used?

like `dispatchExecutor` used [here](https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/dispatcher/DispatchManager.java#L159)
and rgroup.submit happens [here](https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/dispatcher/DispatchManager.java#L223):

Also,  [here](https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/connector/CatalogPruneTask.java#L128), where the `dispatchExecutor` (unbounded) is also used to submit the queueExecution

but that would be protected by the first [call]((https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/dispatcher/DispatchManager.java#L159)).

Reply:
> I might be wrong, but with the resource group, we have seen that even though there were max queue defined, it was not blocking the queries.

According to doc:
```
maxQueued (required): maximum number of queued queries. Once this limit is reached new queries are rejected.
```
other queries should be gracefully rejected (desired behavior).